src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SqlCommand.cs

```diff
@@ -93,7 +84,31 @@ protected override void Clear()
 
             protected override void AfterCleared(SqlCommand owner)
             {
-            
+                owner?.SetCachedCommandExecuteNonQueryAsyncContext(this);
+            }
+        }
+
+        internal sealed class ExecuteXmlReaderAsyncCallContext : AAsyncCallContext<SqlCommand, XmlReader, CancellationTokenRegistration>
```

Comment:
I considered this but I think it takes the type inheritance from 3 levels to 4 and it felt a bit silly. Since every call site either uses the least or most derived types adding things in the middle didn't add anything but cognitive load to understand it all. If there's a technical benefit to doing so then I'm fine with it but without it just seemed like OO for the sake of it.

Reply:
Just considering keeping code reusable as much as possible. Not trivial though.

Reply:
Not too hard and I'm not against doing it but it isn't needed for the current functionality. If we end up with a third (or more) using that same pattern then it would be a clear win and i'd probably do it.